geo: fix GeoJSON and allow options in ST_AsGeoJSON

otan opened the pull request:
There was a bug in GeoJSON where we used the high level name:"Feature"
field, as opposed to the inner Geometry. This has been fixed.

We also support more options for ST_AsGeoJSON now that these changes are
in twpayne/go-geom.

Resolves https://github.com/cockroachdb/cockroach/issues/48381.

Release note (sql change): Implement ST_AsGeoJSON with options to show
bbox and CRS information.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49888)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49888)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/encode_test.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49888#-M9Is0b43l7XjN0i5qjB:-M9Is0b43l7XjN0i5qjC:b-9gxuzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/59292a9c3f4232ef05ed4b09a8af318043777a2c/pkg/geo/encode_test.go#L99)):*
> ```Go
> 		{"SRID=4004;POINT(1.0 1.0)", EWKBToGeoJSONFlagShortCRS, `{"type":"Point","crs":{"type":"name","properties":{"name":"EPSG:4004"}},"coordinates":[1,1]}`},
> 		{"SRID=4004;POINT(1.0 1.0)", EWKBToGeoJSONFlagShortCRS | EWKBToGeoJSONFlagIncludeBBox, `{"type":"Point","bbox":[1,1,1,1],"crs":{"type":"name","properties":{"name":"EPSG:4004"}},"coordinates":[1,1]}`},
> 		{"SRID=4326;POINT(1.0 1.0)", EWKBToGeoJSONFlagShortCRSIfNot4326, `{"type":"Point","coordinates":[1,1]}`},
> ```

can you repeat this last case with "SRID=4004"
___
*[pkg/geo/parse_test.go, line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49888#-M9IsQxU2W8qn3MsDZzP:-M9IsQxU2W8qn3MsDZzQ:bgqjhi4) ([raw file](https://github.com/cockroachdb/cockroach/blob/59292a9c3f4232ef05ed4b09a8af318043777a2c/pkg/geo/parse_test.go#L228)):*
> ```Go
> 		},
> 		{
> 			`{ "type": "Point", "coordinates": [1.0, 1.0] }`,
> ```

I'm assuming we don't currently support the following "feature" example from the PostGIS documentation -- correct?
```
SELECT ST_AsGeoJSON(t.*)
FROM (VALUES
  (1, 'one', 'POINT(1 1)'::geometry),
  (2, 'two', 'POINT(2 2)'),
  (3, 'three', 'POINT(3 3)'))
AS t(id, name, geom);
                                                  st_asgeojson
-----------------------------------------------------------------------------------------------------------------
 {"type": "Feature", "geometry": {"type":"Point","coordinates":[1,1]}, "properties": {"id": 1, "name": "one"}}
 {"type": "Feature", "geometry": {"type":"Point","coordinates":[2,2]}, "properties": {"id": 2, "name": "two"}}
 {"type": "Feature", "geometry": {"type":"Point","coordinates":[3,3]}, "properties": {"id": 3, "name": "three"}}
```
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 948 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49888#-M9ItNqf0683_WustCyB:-M9ItNqf0683_WustCyC:b-e484wb) ([raw file](https://github.com/cockroachdb/cockroach/blob/59292a9c3f4232ef05ed4b09a8af318043777a2c/pkg/sql/sem/builtins/geo_builtins.go#L948)):*
> ```Go
> 
> Options is a flag that can be bitmasked. The options are:
> * 0: no option
> ```

https://postgis.net/docs/ST_AsGeoJSON.html suggests 0 is the default for geography
`text ST_AsGeoJSON(geography geog, integer maxdecimaldigits=9, integer options=0);`


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49888)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/encode_test.go, line 99 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49888#-M9Is0b43l7XjN0i5qjB:-M9JLQjb5O-oskQIdIbm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/59292a9c3f4232ef05ed4b09a8af318043777a2c/pkg/geo/encode_test.go#L99)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you repeat this last case with "SRID=4004"
</blockquote></details>

Done.
___
*[pkg/geo/parse_test.go, line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49888#-M9IsQxU2W8qn3MsDZzP:-M9JKtv40adr-9Wxe_79:bvq695e) ([raw file](https://github.com/cockroachdb/cockroach/blob/59292a9c3f4232ef05ed4b09a8af318043777a2c/pkg/geo/parse_test.go#L228)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm assuming we don't currently support the following "feature" example from the PostGIS documentation -- correct?
```
SELECT ST_AsGeoJSON(t.*)
FROM (VALUES
  (1, 'one', 'POINT(1 1)'::geometry),
  (2, 'two', 'POINT(2 2)'),
  (3, 'three', 'POINT(3 3)'))
AS t(id, name, geom);
                                                  st_asgeojson
-----------------------------------------------------------------------------------------------------------------
 {"type": "Feature", "geometry": {"type":"Point","coordinates":[1,1]}, "properties": {"id": 1, "name": "one"}}
 {"type": "Feature", "geometry": {"type":"Point","coordinates":[2,2]}, "properties": {"id": 2, "name": "two"}}
 {"type": "Feature", "geometry": {"type":"Point","coordinates":[3,3]}, "properties": {"id": 3, "name": "three"}}
```
</blockquote></details>

Yeah we don't suppoer the "record" type.
___
*[pkg/sql/sem/builtins/geo_builtins.go, line 948 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49888#-M9ItNqf0683_WustCyB:-M9JLdM59dnUUvr1JfLu:brnaucg) ([raw file](https://github.com/cockroachdb/cockroach/blob/59292a9c3f4232ef05ed4b09a8af318043777a2c/pkg/sql/sem/builtins/geo_builtins.go#L948)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

https://postgis.net/docs/ST_AsGeoJSON.html suggests 0 is the default for geography
`text ST_AsGeoJSON(geography geog, integer maxdecimaldigits=9, integer options=0);`
</blockquote></details>

oops, typo


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49888#-:-M9JjDwR7w3b_n5fekpF:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola

otan commented:
bors r-

otan committed:
geo: fix GeoJSON and allow options in ST_AsGeoJSON

There was a bug in GeoJSON where we used the high level name:"Feature"
field, as opposed to the inner Geometry. This has been fixed.

We also support more options for ST_AsGeoJSON now that these changes are
in twpayne/go-geom.

Release note (sql change): Implement ST_AsGeoJSON with options to show
bbox and CRS information.

otan commented:
bors r=sumeerbhola
